packages/vite/src/node/preview.ts

```diff
@@ -129,16 +132,59 @@ export async function preview(
       }
     },
   })
-  app.use(previewBase, async (req, res, next) => {
-    if (shouldServe(req.url!, distDir)) {
-      return assetServer(req, res, next)
+  app.use(
+    previewBase,
+    // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
+    async function vitePreviewStaticMiddleware(req, res, next) {
+      if (shouldServe(req.url!, distDir)) {
+        return assetServer(req, res, next)
+      }
+      next()
+    },
+  )
+
+  // html fallback
+  if (config.appType === 'spa' || config.appType === 'mpa') {
+    // append trailing slash when base didn't have it
+    if (config.rawBase !== config.base) {
+      // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
+      app.use(function viteRewriteBaseAccessWithoutTrailingSlashMiddleware(
+        req,
+        res,
+        next,
+      ) {
+        try {
+          const pathname = decodeURIComponent(
+            new URL(req.url!, 'http://example.com').pathname,
+          )
+          if (pathname === config.rawBase) {
+            req.url = config.base + req.url!.slice(config.rawBase.length)
+          }
+        } catch {}
+        next()
+      })
     }
-    next()
-  })
+
+    app.use(
```

Comment:
dev only uses this if `spa` or `mpa`. we'd probably want to do the same here

https://github.com/vitejs/vite/blob/feb8ce05bf2c07f9ea1c6facc75248d2c0e995b8/packages/vite/src/node/server/index.ts#L589

Reply:
This is already inside that condition (line 147).